File: FirmwareSource/__backup/libraries - Kopie/CmdMessenger/CSharp/CommandMessenger/SendCommand.cs
#region CmdMessenger - MIT - (c) 2013 Thijs Elenbaas.
/*
  CmdMessenger - library that provides command based messaging

  Permission is hereby granted, free of charge, to any person obtaining
  a copy of this software and associated documentation files (the
  "Software"), to deal in the Software without restriction, including
  without limitation the rights to use, copy, modify, merge, publish,
  distribute, sublicense, and/or sell copies of the Software, and to
  permit persons to whom the Software is furnished to do so, subject to
  the following conditions:

  The above copyright notice and this permission notice shall be
  included in all copies or substantial portions of the Software.

  Copyright 2013 - Thijs Elenbaas
*/
#endregion

using System;
using System.Collections.Generic;
using System.Globalization;

namespace CommandMessenger
{
    /// <summary> A command to be send by CmdMessenger </summary>
    public class SendCommand : Command
    {
        /// <summary> Indicates if we want to wait for an acknowlegde command. </summary>
        /// <value> true if request acknowledge, false if not. </value>
        public bool ReqAc { get; set; }

        /// <summary> Gets or sets the acknowledge command ID. </summary>
        /// <value> the acknowledge command ID. </value>
        public int AckCmdId { get; set; }

        /// <summary> Gets or sets the time we want to wait for the acknowledgde command. </summary>
        /// <value> The timeout on waiting for an acknowlegde</value>
        public int Timeout { get; set; }

        /// <summary> Constructor. </summary>
        /// <param name="cmdId"> the command ID. </param>
        public SendCommand(int cmdId)
        {
            Init(cmdId, false, 0, 0);
        }

        /// <summary> Constructor. </summary>
        /// <param name="cmdId">    Command ID </param>
        /// <param name="argument"> The argument. </param>
        public SendCommand(int cmdId, string argument)
        {
            Init(cmdId, false, 0, 0);
            AddArgument(argument);
        }

        /// <summary> Constructor. </summary>
        /// <param name="cmdId">     Command ID </param>
        /// <param name="arguments"> The arguments. </param>
        public SendCommand(int cmdId, string[] arguments)
        {
            Init(cmdId, false, 0, 0);
            AddArguments(arguments);
        }

        /// <summary> Constructor. </summary>
        /// <param name="cmdId">    Command ID </param>
        /// <param name="argument"> The argument. </param>
        public SendCommand(int cmdId, float argument)
        {
            Init(cmdId, false, 0, 0);
            AddArgument(argument);
        }

        /// <summary> Constructor. </summary>
        /// <param name="cmdId">    Command ID </param>
        /// <param name="argument"> The argument. </param>
        public SendCommand(int cmdId, double argument)
        {
            Init(cmdId, false, 0, 0);
            AddArgument(argument);
        }

        /// <summary> Constructor. </summary>
        /// <param name="cmdId">    Command ID </param>
        /// <param name="argument"> The argument. </param>
        public SendCommand(int cmdId, UInt16 argument)
        {
            Init(cmdId, false, 0, 0);
            AddArgument(argument);
        }

        /// <summary> Constructor. </summary>
        /// <param name="cmdId">    Command ID </param>
        /// <param name="argument"> The argument. </param>
        public SendCommand(int cmdId, Int16 argument)
        {
            Init(cmdId, false, 0, 0);
            AddArgument(argument);
        }

        /// <summary> Constructor. </summary>
        /// <param name="cmdId">    Command ID </param>
        /// <param name="argument"> The argument. </param>
        public SendCommand(int cmdId, UInt32 argument)
        {
            Init(cmdId, false, 0, 0);
            AddArgument(argument);
        }

        /// <summary> Constructor. </summary>
        /// <param name="cmdId">    Command ID </param>
        /// <param name="argument"> The argument. </param>
        public SendCommand(int cmdId, Int32 argument)
        {
            Init(cmdId, false, 0, 0);
            AddArgument(argument);
        }

        /// <summary> Constructor. </summary>
        /// <param name="cmdId">    Command ID </param>
        /// <param name="argument"> The argument. </param>
        public SendCommand(int cmdId, bool argument)
        {
            Init(cmdId, false, 0, 0);
            AddArgument(argument);
        }

        /// <summary> Constructor. </summary>
        /// <param name="cmdId">    Command ID </param>
        /// <param name="ackCmdId"> Acknowlegde command ID. </param>
        /// <param name="timeout">  The timeout on waiting for an acknowlegde</param>
        public SendCommand(int cmdId, int ackCmdId, int timeout)
        {
            Init(cmdId, true, ackCmdId, timeout);
        }

        /// <summary> Constructor. </summary>
        /// <param name="cmdId">    Command ID </param>
        /// <param name="argument"> The argument. </param>
        /// <param name="ackCmdId"> Acknowlegde command ID. </param>
        /// <param name="timeout">  The timeout on waiting for an acknowlegde</param>
        public SendCommand(int cmdId, string argument, int ackCmdId, int timeout)
        {
            Init(cmdId, true, ackCmdId, timeout);
            AddArgument(argument);
        }

        /// <summary> Constructor. </summary>
        /// <param name="cmdId">     Command ID </param>
        /// <param name="arguments"> The arguments. </param>
        /// <param name="ackCmdId">  Acknowlegde command ID. </param>
        /// <param name="timeout">   The timeout on waiting for an acknowlegde</param>
        public SendCommand(int cmdId, string[] arguments, int ackCmdId, int timeout)
        {
            Init(cmdId, true, ackCmdId, timeout);
            AddArguments(arguments);
        }

        /// <summary> Constructor. </summary>
        /// <param name="cmdId">    Command ID </param>
        /// <param name="argument"> The argument. </param>
        /// <param name="ackCmdId"> Acknowlegde command ID. </param>
        /// <param name="timeout">  The timeout on waiting for an acknowlegde</param>
        public SendCommand(int cmdId, float argument, int ackCmdId, int timeout)
        {
            Init(cmdId, true, ackCmdId, timeout);
            AddArgument(argument);
        }

        /// <summary> Constructor. </summary>
        /// <param name="cmdId">    Command ID </param>
        /// <param name="argument"> The argument. </param>
        /// <param name="ackCmdId"> Acknowlegde command ID. </param>
        /// <param name="timeout">  The timeout on waiting for an acknowlegde</param>
        public SendCommand(int cmdId, double argument, int ackCmdId, int timeout)
        {
            Init(cmdId, true, ackCmdId, timeout);
            AddArgument(argument);
        }

        /// <summary> Constructor. </summary>
        /// <param name="cmdId">    Command ID </param>
        /// <param name="argument"> The argument. </param>
        /// <param name="ackCmdId"> Acknowlegde command ID. </param>
        /// <param name="timeout">  The timeout on waiting for an acknowlegde</param>
        public SendCommand(int cmdId, Int16 argument, int ackCmdId, int timeout)
        {
            Init(cmdId, true, ackCmdId, timeout);
            AddArgument(argument);
        }

        /// <summary> Constructor. </summary>
        /// <param name="cmdId">    Command ID </param>
        /// <param name="argument"> The argument. </param>
        /// <param name="ackCmdId"> Acknowlegde command ID. </param>
        /// <param name="timeout">  The timeout on waiting for an acknowlegde</param>
        public SendCommand(int cmdId, UInt16 argument, int ackCmdId, int timeout)
        {
            Init(cmdId, true, ackCmdId, timeout);
            AddArgument(argument);
        }

        /// <summary> Constructor. </summary>
        /// <param name="cmdId">    Command ID </param>
        /// <param name="argument"> The argument. </param>
        /// <param name="ackCmdId"> Acknowlegde command ID. </param>
        /// <param name="timeout">  The timeout on waiting for an acknowlegde</param>
        public SendCommand(int cmdId, Int32 argument, int ackCmdId, int timeout)
        {
            Init(cmdId, true, ackCmdId, timeout);
            AddArgument(argument);
        }

        /// <summary> Constructor. </summary>
        /// <param name="cmdId">    Command ID </param>
        /// <param name="argument"> The argument. </param>
        /// <param name="ackCmdId"> Acknowlegde command ID. </param>
        /// <param name="timeout">  The timeout on waiting for an acknowlegde</param>
        public SendCommand(int cmdId, UInt32 argument, int ackCmdId, int timeout)
        {
            Init(cmdId, true, ackCmdId, timeout);
            AddArgument(argument);
        }

        /// <summary> Initialises this object. </summary>
        /// <param name="cmdId">    Command ID </param>
        /// <param name="reqAc">    true to request ac. </param>
        /// <param name="ackCmdId"> Acknowlegde command ID. </param>
        /// <param name="timeout">  The timeout on waiting for an acknowlegde</param>
        private void Init(int cmdId, bool reqAc, int ackCmdId, int timeout)
        {
            ReqAc = reqAc;
            CmdId = cmdId;
            AckCmdId = ackCmdId;
            Timeout = timeout;
            _arguments = new List<string>();
        }

        // ***** String based **** /

        /// <summary> Adds a command argument.  </summary>
        /// <param name="argument"> The argument. </param>
        public void AddArgument(string argument)
        {
            if (argument != null)
                _arguments.Add(argument);
        }

        /// <summary> Adds command arguments.  </summary>
        /// <param name="arguments"> The arguments. </param>
        public void AddArguments(string[] arguments)
        {
            if (arguments != null)
                _arguments.AddRange(arguments);
        }

        /// <summary> Adds a command argument. </summary>
        /// <param name="argument"> The argument. </param>
        public void AddArgument(Single argument)
        {
            _arguments.Add(argument.ToString(CultureInfo.InvariantCulture));
        }

        /// <summary> Adds a command argument. </summary>
        /// <param name="argument"> The argument. </param>
        public void AddArgument(Double argument)
        {
            _arguments.Add(argument.ToString(CultureInfo.InvariantCulture));
        }

        /// <summary> Adds a command argument. </summary>
        /// <param name="argument"> The argument. </param>
        public void AddArgument(Int16 argument)
        {
            _arguments.Add(argument.ToString(CultureInfo.InvariantCulture));
        }

        /// <summary> Adds a command argument. </summary>
        /// <param name="argument"> The argument. </param>
        public void AddArgument(UInt16 argument)
        {
            _arguments.Add(argument.ToString(CultureInfo.InvariantCulture));
        }

        /// <summary> Adds a command argument. </summary>
        /// <param name="argument"> The argument. </param>
        public void AddArgument(Int32 argument)
        {
            _arguments.Add(argument.ToString(CultureInfo.InvariantCulture));
        }

        /// <summary> Adds a command argument. </summary>
        /// <param name="argument"> The argument. </param>
        public void AddArgument(UInt32 argument)
        {
            _arguments.Add(argument.ToString(CultureInfo.InvariantCulture));
        }

        /// <summary> Adds a command argument. </summary>
        /// <param name="argument"> The argument. </param>
        public void AddArgument(bool argument)
        {
            AddArgument((Int32) (argument ? 1 : 0));
        }

        // ***** Binary **** /

        /// <summary> Adds a binary command argument.  </summary>
        /// <param name="argument"> The argument. </param>
        public void AddBinArgument(string argument)
        {
            _arguments.Add(Escaping.Escape(argument));
        }

        /// <summary> Adds a binary command argument. </summary>
        /// <param name="argument"> The argument. </param>
        public void AddBinArgument(Single argument)
        {
            _arguments.Add(BinaryConverter.ToString(argument));
        }

        /// <summary> Adds a binary command argument. </summary>
        /// <param name="argument"> The argument. </param>
        public void AddBinArgument(Double argument)
        {
            _arguments.Add(BinaryConverter.ToString(argument));
        }

        /// <summary> Adds a binary command argument. </summary>
        /// <param name="argument"> The argument. </param>
        public void AddBinArgument(Int16 argument)
        {
            _arguments.Add(BinaryConverter.ToString(argument));
        }

        /// <summary> Adds a binary command argument. </summary>
        /// <param name="argument"> The argument. </param>
        public void AddBinArgument(UInt16 argument)
        {
            _arguments.Add(BinaryConverter.ToString(argument));
        }

        /// <summary> Adds a binary command argument. </summary>
        /// <param name="argument"> The argument. </param>
        public void AddBinArgument(Int32 argument)
        {
            _arguments.Add(BinaryConverter.ToString(argument));
        }

        /// <summary> Adds a binary command argument. </summary>
        /// <param name="argument"> The argument. </param>
        public void AddBinArgument(UInt32 argument)
        {
            _arguments.Add(BinaryConverter.ToString(argument));
        }

        /// <summary> Adds a binary command argument. </summary>
        /// <param name="argument"> The argument. </param>
        public void AddBinArgument(bool argument)
        {
            _arguments.Add(BinaryConverter.ToString(argument ? (byte) 1 : (byte) 0));
        }
    }
}

File: FirmwareSource/lib/CmdMessenger/extras/VisualBasic/CommandMessenger/SendCommand.vb
#Region "CmdMessenger - MIT - (c) 2013 Thijs Elenbaas."
'
'  CmdMessenger - library that provides command based messaging
'
'  Permission is hereby granted, free of charge, to any person obtaining
'  a copy of this software and associated documentation files (the
'  "Software"), to deal in the Software without restriction, including
'  without limitation the rights to use, copy, modify, merge, publish,
'  distribute, sublicense, and/or sell copies of the Software, and to
'  permit persons to whom the Software is furnished to do so, subject to
'  the following conditions:
'
'  The above copyright notice and this permission notice shall be
'  included in all copies or substantial portions of the Software.
'
'  Copyright 2013 - Thijs Elenbaas
'
#End Region


Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Globalization

Namespace CommandMessenger
	''' <summary> A command to be send by CmdMessenger </summary>
	Public Class SendCommand
		Inherits Command
		''' <summary> Indicates if we want to wait for an acknowledge command. </summary>
		''' <value> true if request acknowledge, false if not. </value>
		Private privateReqAc As Boolean
		Public Property ReqAc() As Boolean
			Get
				Return privateReqAc
			End Get
			Set(ByVal value As Boolean)
				privateReqAc = value
			End Set
		End Property

		''' <summary> Gets or sets the acknowledge command ID. </summary>
		''' <value> the acknowledge command ID. </value>
		Private privateAckCmdId As Integer
		Public Property AckCmdId() As Integer
			Get
				Return privateAckCmdId
			End Get
			Set(ByVal value As Integer)
				privateAckCmdId = value
			End Set
		End Property

		''' <summary> Gets or sets the time we want to wait for the acknowledge command. </summary>
		''' <value> The timeout on waiting for an acknowledge</value>
		Private privateTimeout As Integer
		Public Property Timeout() As Integer
			Get
				Return privateTimeout
			End Get
			Set(ByVal value As Integer)
				privateTimeout = value
			End Set
		End Property

		''' <summary> Constructor. </summary>
		''' <param name="cmdId"> the command ID. </param>
		Public Sub New(ByVal cmdId As Integer)
			Init(cmdId, False, 0, 0)
		End Sub

		''' <summary> Constructor. </summary>
		''' <param name="cmdId">    Command ID </param>
		''' <param name="argument"> The argument. </param>
		Public Sub New(ByVal cmdId As Integer, ByVal argument As String)
			Init(cmdId, False, 0, 0)
			AddArgument(argument)
		End Sub

		''' <summary> Constructor. </summary>
		''' <param name="cmdId">     Command ID </param>
		''' <param name="arguments"> The arguments. </param>
		Public Sub New(ByVal cmdId As Integer, ByVal arguments() As String)
			Init(cmdId, False, 0, 0)
			AddArguments(arguments)
		End Sub

		''' <summary> Constructor. </summary>
		''' <param name="cmdId">    Command ID </param>
		''' <param name="argument"> The argument. </param>
		Public Sub New(ByVal cmdId As Integer, ByVal argument As Single)
			Init(cmdId, False, 0, 0)
			AddArgument(argument)
		End Sub

		''' <summary> Constructor. </summary>
		''' <param name="cmdId">    Command ID </param>
		''' <param name="argument"> The argument. </param>
		Public Sub New(ByVal cmdId As Integer, ByVal argument As Double)
			Init(cmdId, False, 0, 0)
			AddArgument(argument)
		End Sub

		''' <summary> Constructor. </summary>
		''' <param name="cmdId">    Command ID </param>
		''' <param name="argument"> The argument. </param>
		Public Sub New(ByVal cmdId As Integer, ByVal argument As UInt16)
			Init(cmdId, False, 0, 0)
			AddArgument(argument)
		End Sub

		''' <summary> Constructor. </summary>
		''' <param name="cmdId">    Command ID </param>
		''' <param name="argument"> The argument. </param>
		Public Sub New(ByVal cmdId As Integer, ByVal argument As Int16)
			Init(cmdId, False, 0, 0)
			AddArgument(argument)
		End Sub

		''' <summary> Constructor. </summary>
		''' <param name="cmdId">    Command ID </param>
		''' <param name="argument"> The argument. </param>
		Public Sub New(ByVal cmdId As Integer, ByVal argument As UInt32)
			Init(cmdId, False, 0, 0)
			AddArgument(argument)
		End Sub

		''' <summary> Constructor. </summary>
		''' <param name="cmdId">    Command ID </param>
		''' <param name="argument"> The argument. </param>
		Public Sub New(ByVal cmdId As Integer, ByVal argument As Int32)
			Init(cmdId, False, 0, 0)
			AddArgument(argument)
		End Sub

		''' <summary> Constructor. </summary>
		''' <param name="cmdId">    Command ID </param>
		''' <param name="argument"> The argument. </param>
		Public Sub New(ByVal cmdId As Integer, ByVal argument As Boolean)
			Init(cmdId, False, 0, 0)
			AddArgument(argument)
		End Sub

		''' <summary> Constructor. </summary>
		''' <param name="cmdId">    Command ID </param>
		''' <param name="ackCmdId"> Acknowledge command ID. </param>
		''' <param name="timeout">  The timeout on waiting for an acknowledge</param>
		Public Sub New(ByVal cmdId As Integer, ByVal ackCmdId As Integer, ByVal timeout As Integer)
			Init(cmdId, True, ackCmdId, timeout)
		End Sub

		''' <summary> Constructor. </summary>
		''' <param name="cmdId">    Command ID </param>
		''' <param name="argument"> The argument. </param>
		''' <param name="ackCmdId"> Acknowledge command ID. </param>
		''' <param name="timeout">  The timeout on waiting for an acknowledge</param>
		Public Sub New(ByVal cmdId As Integer, ByVal argument As String, ByVal ackCmdId As Integer, ByVal timeout As Integer)
			Init(cmdId, True, ackCmdId, timeout)
			AddArgument(argument)
		End Sub

		''' <summary> Constructor. </summary>
		''' <param name="cmdId">     Command ID </param>
		''' <param name="arguments"> The arguments. </param>
		''' <param name="ackCmdId">  Acknowledge command ID. </param>
		''' <param name="timeout">   The timeout on waiting for an acknowledge</param>
		Public Sub New(ByVal cmdId As Integer, ByVal arguments() As String, ByVal ackCmdId As Integer, ByVal timeout As Integer)
			Init(cmdId, True, ackCmdId, timeout)
			AddArguments(arguments)
		End Sub

		''' <summary> Constructor. </summary>
		''' <param name="cmdId">    Command ID </param>
		''' <param name="argument"> The argument. </param>
		''' <param name="ackCmdId"> Acknowledge command ID. </param>
		''' <param name="timeout">  The timeout on waiting for an acknowledge</param>
		Public Sub New(ByVal cmdId As Integer, ByVal argument As Single, ByVal ackCmdId As Integer, ByVal timeout As Integer)
			Init(cmdId, True, ackCmdId, timeout)
			AddArgument(argument)
		End Sub

		''' <summary> Constructor. </summary>
		''' <param name="cmdId">    Command ID </param>
		''' <param name="argument"> The argument. </param>
		''' <param name="ackCmdId"> Acknowledge command ID. </param>
		''' <param name="timeout">  The timeout on waiting for an acknowledge</param>
		Public Sub New(ByVal cmdId As Integer, ByVal argument As Double, ByVal ackCmdId As Integer, ByVal timeout As Integer)
			Init(cmdId, True, ackCmdId, timeout)
			AddArgument(argument)
		End Sub

		''' <summary> Constructor. </summary>
		''' <param name="cmdId">    Command ID </param>
		''' <param name="argument"> The argument. </param>
		''' <param name="ackCmdId"> Acknowledge command ID. </param>
		''' <param name="timeout">  The timeout on waiting for an acknowledge</param>
		Public Sub New(ByVal cmdId As Integer, ByVal argument As Int16, ByVal ackCmdId As Integer, ByVal timeout As Integer)
			Init(cmdId, True, ackCmdId, timeout)
			AddArgument(argument)
		End Sub

		''' <summary> Constructor. </summary>
		''' <param name="cmdId">    Command ID </param>
		''' <param name="argument"> The argument. </param>
		''' <param name="ackCmdId"> Acknowledge command ID. </param>
		''' <param name="timeout">  The timeout on waiting for an acknowledge</param>
		Public Sub New(ByVal cmdId As Integer, ByVal argument As UInt16, ByVal ackCmdId As Integer, ByVal timeout As Integer)
			Init(cmdId, True, ackCmdId, timeout)
			AddArgument(argument)
		End Sub

		''' <summary> Constructor. </summary>
		''' <param name="cmdId">    Command ID </param>
		''' <param name="argument"> The argument. </param>
		''' <param name="ackCmdId"> Acknowledge command ID. </param>
		''' <param name="timeout">  The timeout on waiting for an acknowledge</param>
		Public Sub New(ByVal cmdId As Integer, ByVal argument As Int32, ByVal ackCmdId As Integer, ByVal timeout As Integer)
			Init(cmdId, True, ackCmdId, timeout)
			AddArgument(argument)
		End Sub

		''' <summary> Constructor. </summary>
		''' <param name="cmdId">    Command ID </param>
		''' <param name="argument"> The argument. </param>
		''' <param name="ackCmdId"> Acknowledge command ID. </param>
		''' <param name="timeout">  The timeout on waiting for an acknowledge</param>
		Public Sub New(ByVal cmdId As Integer, ByVal argument As UInt32, ByVal ackCmdId As Integer, ByVal timeout As Integer)
			Init(cmdId, True, ackCmdId, timeout)
			AddArgument(argument)
		End Sub

		''' <summary> Initializes this object. </summary>
		''' <param name="cmdId">    Command ID </param>
		''' <param name="reqAc">    true to request ac. </param>
		''' <param name="ackCmdId"> Acknowledge command ID. </param>
		''' <param name="timeout">  The timeout on waiting for an acknowledge</param>
		Private Sub Init(ByVal cmdId As Integer, ByVal reqAc As Boolean, ByVal ackCmdId As Integer, ByVal timeout As Integer)
			Me.ReqAc = reqAc
			Me.CmdId = cmdId
			Me.AckCmdId = ackCmdId
			Me.Timeout = timeout
			_arguments = New List(Of String)()
		End Sub

		' ***** String based **** /

		''' <summary> Adds a command argument.  </summary>
		''' <param name="argument"> The argument. </param>
		Public Sub AddArgument(ByVal argument As String)
			If argument IsNot Nothing Then
				_arguments.Add(argument)
			End If
		End Sub

		''' <summary> Adds command arguments.  </summary>
		''' <param name="arguments"> The arguments. </param>
		Public Sub AddArguments(ByVal arguments() As String)
			If arguments IsNot Nothing Then
				_arguments.AddRange(arguments)
			End If
		End Sub



		''' <summary> Adds a command argument. </summary>
		''' <param name="argument"> The argument. </param>
		Public Sub AddArgument(ByVal argument As Single)
			_arguments.Add(argument.ToString("R",CultureInfo.InvariantCulture))
		End Sub

		''' <summary> Adds a command argument. </summary>
		''' <param name="argument"> The argument. </param>
		Public Sub AddArgument(ByVal argument As Double)
			If BoardType = BoardType.Bit16 Then
				' Not completely sure if this is needed for plain text sending.
				Dim floatArg = CSng(argument)
				_arguments.Add(floatArg.ToString("R",CultureInfo.InvariantCulture))
			Else
				_arguments.Add(argument.ToString("R",CultureInfo.InvariantCulture))
			End If
		End Sub

		''' <summary> Adds a command argument. </summary>
		''' <param name="argument"> The argument. </param>
		Public Sub AddArgument(ByVal argument As Int16)
			_arguments.Add(argument.ToString(CultureInfo.InvariantCulture))
		End Sub

		''' <summary> Adds a command argument. </summary>
		''' <param name="argument"> The argument. </param>
		Public Sub AddArgument(ByVal argument As UInt16)
			_arguments.Add(argument.ToString(CultureInfo.InvariantCulture))
		End Sub

		''' <summary> Adds a command argument. </summary>
		''' <param name="argument"> The argument. </param>
		Public Sub AddArgument(ByVal argument As Int32)
			' Make sure the other side can read this: on a 16 processor, read as Long
			_arguments.Add(argument.ToString(CultureInfo.InvariantCulture))
		End Sub

		''' <summary> Adds a command argument. </summary>
		''' <param name="argument"> The argument. </param>
		Public Sub AddArgument(ByVal argument As UInt32)
			' Make sure the other side can read this: on a 16 processor, read as Long
			_arguments.Add(argument.ToString(CultureInfo.InvariantCulture))
		End Sub

		''' <summary> Adds a command argument. </summary>
		''' <param name="argument"> The argument. </param>
		Public Sub AddArgument(ByVal argument As Boolean)
			AddArgument(CShort(Fix(If(argument, 1, 0))))
		End Sub

		' ***** Binary **** /

		''' <summary> Adds a binary command argument.  </summary>
		''' <param name="argument"> The argument. </param>
		Public Sub AddBinArgument(ByVal argument As String)
			_arguments.Add(Escaping.Escape(argument))
		End Sub

		''' <summary> Adds a binary command argument. </summary>
		''' <param name="argument"> The argument. </param>
		Public Sub AddBinArgument(ByVal argument As Single)
			_arguments.Add(BinaryConverter.ToString(argument))
		End Sub

		''' <summary> Adds a binary command argument. </summary>
		''' <param name="argument"> The argument. </param>
		Public Sub AddBinArgument(ByVal argument As Double)
			_arguments.Add(If(BoardType = BoardType.Bit16, BinaryConverter.ToString(CSng(argument)), BinaryConverter.ToString(argument)))
		End Sub

		''' <summary> Adds a binary command argument. </summary>
		''' <param name="argument"> The argument. </param>
		Public Sub AddBinArgument(ByVal argument As Int16)
			_arguments.Add(BinaryConverter.ToString(argument))
		End Sub

		''' <summary> Adds a binary command argument. </summary>
		''' <param name="argument"> The argument. </param>
		Public Sub AddBinArgument(ByVal argument As UInt16)
			_arguments.Add(BinaryConverter.ToString(argument))
		End Sub

		''' <summary> Adds a binary command argument. </summary>
		''' <param name="argument"> The argument. </param>
		Public Sub AddBinArgument(ByVal argument As Int32)
			_arguments.Add(BinaryConverter.ToString(argument))
		End Sub

		''' <summary> Adds a binary command argument. </summary>
		''' <param name="argument"> The argument. </param>
		Public Sub AddBinArgument(ByVal argument As UInt32)
			_arguments.Add(BinaryConverter.ToString(argument))
		End Sub

		''' <summary> Adds a binary command argument. </summary>
		''' <param name="argument"> The argument. </param>
		Public Sub AddBinArgument(ByVal argument As Boolean)
			_arguments.Add(BinaryConverter.ToString(If(argument, CByte(1), CByte(0))))
		End Sub
	End Class
End Namespace
